File: backend/internal/interface/http/handlers/audit.go
// This file is the org-scoped audit-log viewer + CSV export (FR-AUD-003). Both
// routes are ADMIN-only (read:audit). The usecase forces the org_id filter, the
// isolation backstop for a non-RLS table. CSV is bounded at audit.ExportCap (10k).
package handlers

import (
	"encoding/csv"
	"encoding/json"
	"log/slog"
	"net/http"
	"time"

	"github.com/mesutokul/fluxboard/backend/internal/domain"
	mw "github.com/mesutokul/fluxboard/backend/internal/interface/http/middleware"
	"github.com/mesutokul/fluxboard/backend/internal/interface/http/response"
	"github.com/mesutokul/fluxboard/backend/internal/usecase/audituc"
)

// AuditHandlers serves the org audit viewer.
type AuditHandlers struct {
	svc *audituc.Service
	log *slog.Logger
}

// NewAuditHandlers builds AuditHandlers.
func NewAuditHandlers(svc *audituc.Service, log *slog.Logger) *AuditHandlers {
	if log == nil {
		log = slog.Default()
	}
	return &AuditHandlers{svc: svc, log: log}
}

// List: GET /orgs/{orgId}/audit?actor=&action=&severity=&since=&until=&limit=
func (h *AuditHandlers) List(w http.ResponseWriter, r *http.Request) {
	tc, ok := mw.TenantFrom(r.Context())
	if !ok {
		response.Error(w, domain.ErrForbidden)
		return
	}
	entries, err := h.svc.ListForOrg(r.Context(), tc.OrgID, auditFilterFromQuery(r), false)
	if err != nil {
		response.Error(w, err)
		return
	}
	response.JSON(w, http.StatusOK, map[string]any{"entries": toAuditEntries(entries)})
}

// ExportCSV: GET /orgs/{orgId}/audit.csv — streams up to ExportCap rows.
func (h *AuditHandlers) ExportCSV(w http.ResponseWriter, r *http.Request) {
	tc, ok := mw.TenantFrom(r.Context())
	if !ok {
		response.Error(w, domain.ErrForbidden)
		return
	}
	entries, err := h.svc.ListForOrg(r.Context(), tc.OrgID, auditFilterFromQuery(r), true)
	if err != nil {
		response.Error(w, err)
		return
	}
	w.Header().Set("Content-Type", "text/csv; charset=utf-8")
	w.Header().Set("Content-Disposition", `attachment; filename="audit.csv"`)
	cw := csv.NewWriter(w)
	_ = cw.Write([]string{"created_at", "action", "severity", "actor_user_id",
		"impersonator_user_id", "target_type", "target_id", "ip", "metadata"})
	for _, e := range entries {
		meta := ""
		if len(e.Metadata) > 0 {
			if b, err := json.Marshal(e.Metadata); err == nil {
				meta = string(b)
			}
		}
		_ = cw.Write([]string{
			e.CreatedAt.UTC().Format(time.RFC3339),
			string(e.Action), string(e.Severity), e.ActorUserID, e.ImpersonatorUserID,
			e.TargetType, e.TargetID, e.IP, meta,
		})
	}
	cw.Flush()
}

File: web/lib/api/audit.ts
import { apiFetch } from './client';
import type { AuditEntry } from './types';

// Org audit-log viewer + CSV export (docs/08 §6, FR-AUD-003). Both ADMIN-only
// (read:audit); the usecase forces the org_id filter server-side.

export interface AuditFilter {
  actor?: string;
  action?: string;
  severity?: string;
  /** RFC3339 lower bound. */
  since?: string;
  /** RFC3339 upper bound. */
  until?: string;
  limit?: number;
}

export function toQuery(filter: AuditFilter): string {
  const qs = new URLSearchParams();
  for (const [k, v] of Object.entries(filter)) {
    if (v !== undefined && v !== '' && v !== null) qs.set(k, String(v));
  }
  const s = qs.toString();
  return s ? `?${s}` : '';
}

export async function listAudit(orgId: string, filter: AuditFilter = {}): Promise<AuditEntry[]> {
  const res = await apiFetch<{ entries: AuditEntry[] }>(`/orgs/${orgId}/audit${toQuery(filter)}`);
  return res.entries;
}

/**
 * Fetch the CSV export as text. apiFetch returns the raw body for non-JSON
 * responses; the caller wraps it in a Blob to trigger a download (a plain
 * `<a download>` can't attach the in-memory bearer token).
 */
export function auditCsv(orgId: string, filter: AuditFilter = {}): Promise<string> {
  return apiFetch<string>(`/orgs/${orgId}/audit.csv${toQuery(filter)}`);
}
